internal/operator-controller/authorization/rbac.go

```diff
@@ -0,0 +1,605 @@
+func (ec *escalationChecker) checkRoleEscalation(ctx context.Context, role *rbacv1.Role) error {
+	if rbacregistry.EscalationAllowed(ctx) || rbacregistry.RoleEscalationAuthorized(ctx, ec.authorizer) {
+		return nil
+	}
+
+	rules := role.Rules
+	if err := validation.ConfirmNoEscalation(ctx, ec.ruleResolver, rules); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (ec *escalationChecker) checkRoleBindingEscalation(ctx context.Context, roleBinding *rbacv1.RoleBinding) error {
+	if rbacregistry.EscalationAllowed(ctx) {
+		return nil
+	}
+
+	roleRef := rbacinternal.RoleRef{}
+	err := rbacv1helpers.Convert_v1_RoleRef_To_rbac_RoleRef(&roleBinding.RoleRef, &roleRef, nil)
+	if err != nil {
+		return err
+	}
+	if rbacregistry.BindingAuthorized(ctx, roleRef, roleBinding.Namespace, ec.authorizer) {
+		return nil
+	}
+
+	rules, err := ec.ruleResolver.GetRoleReferenceRules(ctx, roleBinding.RoleRef, roleBinding.Namespace)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	switch roleRef.Kind {
+	case "ClusterRole":
+		if manifestClusterRole, ok := ec.extraClusterRoles[types.NamespacedName{Name: roleBinding.RoleRef.Name}]; ok {
+			rules = append(rules, manifestClusterRole.Rules...)
+		}
+	case "Role":
+		if manifestRole, ok := ec.extraRoles[types.NamespacedName{Namespace: roleBinding.Namespace, Name: roleBinding.RoleRef.Name}]; ok {
+			rules = append(rules, manifestRole.Rules...)
+		}
+	}
+
+	if err := validation.ConfirmNoEscalation(ctx, ec.ruleResolver, rules); err != nil {
+		return err
+	}
+	return nil
+}
+
+var fullAuthority = []rbacv1.PolicyRule{
+	{Verbs: []string{"*"}, APIGroups: []string{"*"}, Resources: []string{"*"}},
+	{Verbs: []string{"*"}, NonResourceURLs: []string{"*"}},
+}
+
+// TODO: Investigate replacing this regex parsing with structured error handling once there are
+//
+//	structured RBAC errors introduced by https://github.com/kubernetes/kubernetes/pull/130955.
+//
+// parseEscalationErrorForMissingRules attempts to extract specific RBAC permissions
+// that were denied due to escalation prevention from a given error's text.
+// It returns the list of extracted PolicyRules and an error.
+// Note: If parsing is successful, the returned error is derived from the *input* error's
+// message (specifically the part indicating the escalation attempt), not an error
+// encountered during the parsing process itself. If parsing fails due to an unexpected
+// error format, a distinct parsing error is returned.
+func parseEscalationErrorForMissingRules(ecError error) ([]rbacv1.PolicyRule, error) {
+	// errRegex captures the standard prefix of an escalation error message
+	errRegex := regexp.MustCompile(`(?s)^(user ".*" \(groups=.*\) is attempting to grant RBAC permissions not currently held):.*?$`)
+	// permRegex extracts the details (APIGroups, Resources, Verbs) of individual permissions listed within the error message
+	permRegex := regexp.MustCompile(`{APIGroups:\[("[^"]*")], Resources:\[("[^"]*")], Verbs:\[("[^"]*")]}`)
```

Comment:
Does this only work correctly when APIGroups, Resources, and Verbs contain a single item?

It seems like a string like `{APIGroups:[""], Resources:["secrets","configmaps"], Verbs:["get", "list", "watch"]}` will extract `"secrets","configmaps"` and `"get", "list", "watch"` as a single item, and then below we trim the leading and trailing `"`, leaving `secrets","configmaps` and `get", "list", "watch`?

Reply:
I changed this to match any number of items inside the [ ], but I'm not sure if this is irrelevant per my question on your other thread on the error regexp

Reply:
these got done in later commits, resolving